internal/k8s/statusaddress_test.go

```diff
@@ -0,0 +1,355 @@
+	})
+
+}
+
+func TestStatusAddressUpdater_OnUpdate(t *testing.T) {
+
+	ingressGVR := v1beta1.SchemeGroupVersion.WithResource("ingresses")
+	log := logrus.New()
+	log.SetLevel(logrus.DebugLevel)
+	emptyLBStatus := v1.LoadBalancerStatus{}
+
+	IPLBStatus := v1.LoadBalancerStatus{
+		Ingress: []v1.LoadBalancerIngress{
+			{
+				IP: "127.0.0.1",
+			},
+		},
+	}
+
+	type testcase struct {
+		status       v1.LoadBalancerStatus
+		ingressClass string
+		objname      string
+		preop        interface{}
+		postop       interface{}
+	}
+
+	run := func(t *testing.T, name string, tc testcase) {
+		suc := StatusUpdateCacher{}
+		if !suc.AddObject(tc.objname, tc.objname, ingressGVR, tc.preop) {
+			t.Fatal("unable to add object to cache")
+		}
+
+		isu := StatusAddressUpdater{
+			Logger:        log,
+			LBStatus:      tc.status,
+			IngressClass:  tc.ingressClass,
+			StatusUpdater: &suc,
+		}
+
+		isu.OnUpdate(tc.preop, tc.preop)
+
+		newObj := suc.GetObject(tc.objname, tc.objname, ingressGVR)
+		if !assert.StatusEqual(tc.postop, newObj) {
+			t.Fatalf("%s: Status not equal: %s\n", name, cmp.Diff(tc.postop, newObj))
+		}
+
+	}
+
+	run(t, "no-op add", testcase{
+		status:       emptyLBStatus,
+		ingressClass: "",
+		objname:      "noop",
+		preop:        simpleIngressGenerator("noop", "", emptyLBStatus),
+		postop:       simpleIngressGenerator("noop", "", emptyLBStatus),
+	})
+
+	run(t, "add an IP should update", testcase{
+		status:       emptyLBStatus,
+		ingressClass: "",
+		objname:      "withIP",
+		preop:        simpleIngressGenerator("withIP", "", emptyLBStatus),
+		postop:       simpleIngressGenerator("withIP", "", emptyLBStatus),
+	})
+
+	run(t, "unset ingressclass should not update", testcase{
+		status:       IPLBStatus,
+		ingressClass: "phony",
+		objname:      "unsetingressclass",
+		preop:        simpleIngressGenerator("unsetingressclass", "", emptyLBStatus),
+		postop:       simpleIngressGenerator("unsetingressclass", "", emptyLBStatus),
+	})
+
+	run(t, "matching ingressclass should not update", testcase{
+		status:       IPLBStatus,
+		ingressClass: "phony",
+		objname:      "nonmatchingingressclass",
+		preop:        simpleIngressGenerator("nonmatchingingressclass", "other", emptyLBStatus),
+		postop:       simpleIngressGenerator("nonmatchingingressclass", "other", emptyLBStatus),
+	})
+
+	run(t, "matching ingressclass should update", testcase{
+		status:       IPLBStatus,
+		ingressClass: "phony",
+		objname:      "matchingingressclass",
+		preop:        simpleIngressGenerator("matchingingressclass", "phony", emptyLBStatus),
+		postop:       simpleIngressGenerator("matchingingressclass", "phony", IPLBStatus),
+	})
+
+}
+
+func simpleIngressGenerator(name, ingressClass string, lbstatus v1.LoadBalancerStatus) *v1beta1.Ingress {
+	return &v1beta1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: name,
+			Annotations: map[string]string{
+				"kubernetes.io/ingress.class": ingressClass,
```

Comment:
Should we also test using `projectcontour.io/ingress.class`? Might be better as a feature test for that one.

Reply:
We can test using either, that's actually testing the code in `annotation.MatchesIngressClass`, which already has code for `projectcontour.io/ingress.class`. A feature test is a good idea though.,

Reply:
Looks like I was wrong on that one, but @stevesloka is fixing this in #2536, thanks Steve!